lang/java/reef-applications/reef-vortex/src/main/java/org/apache/reef/vortex/api/VortexAggregateFuture.java

```diff
@@ -0,0 +1,223 @@
+      removeCompletedTasklets(result, Collections.singletonList(taskletId));
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Aggregation has completed for a list of Tasklets, with an aggregated result.
+   */
+  @Private
+  @Override
+  public void aggregationCompleted(final List<Integer> pTaskletIds, final byte[] serializedResult) {
+    try {
+      // TODO[REEF-1113]: Handle serialization failure separately in Vortex
+      final TOutput result = aggOutputCodec.decode(serializedResult);
+      removeCompletedTasklets(result, pTaskletIds);
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A Tasklet associated with the aggregation has failed.
+   */
+  @Private
+  @Override
+  public void threwException(final int taskletId, final Exception exception) {
+    try {
+      removeFailedTasklets(exception, Collections.singletonList(taskletId));
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * A list of Tasklets has failed during aggregation phase.
+   */
+  @Private
+  @Override
+  public void aggregationThrewException(final List<Integer> pTaskletIds, final Exception exception) {
+    try {
+      removeFailedTasklets(exception, pTaskletIds);
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Not implemented for local aggregation.
+   */
+  @Private
+  @Override
+  public void cancelled(final int taskletId) {
+    throw new NotImplementedException("Tasklet cancellation not supported in aggregations.");
+  }
+
+  /**
+   * Removes completed Tasklets from Tasklets that are expected and invoke callback.
+   */
+  private synchronized void removeCompletedTasklets(final TOutput output, final List<Integer> taskletIds)
+      throws InterruptedException {
+    final AggregateResult result =
+        new AggregateResult(output, getInputs(taskletIds), taskletIdInputMap.size() > 0);
+
+    if (callbackHandler != null) {
+      executor.execute(new Runnable() {
+        @Override
+        public void run() {
+          callbackHandler.onSuccess(result);
+        }
+      });
+    }
+
+    resultQueue.put(result);
+  }
+
+  /**
+   * Removes failed Tasklets from Tasklets that are expected and invokes callback.
+   */
+  private synchronized void removeFailedTasklets(final Exception exception, final List<Integer> taskletIds)
+      throws InterruptedException {
+
+    final AggregateResult failure =
+        new AggregateResult(exception, getInputs(taskletIds), taskletIdInputMap.size() > 0);
+
+    if (callbackHandler != null) {
+      executor.execute(new Runnable() {
+        @Override
+        public void run() {
+          // TODO[JIRA REEF-1129]: Add documentation in VortexThreadPool.
+          callbackHandler.onFailure(new VortexAggregateException(exception, taskletIds));
```

Comment:
In the `VortexAggregateException`, the second argument is `List<?> inputList` instead of taskletIds. Could you confirm that it is right?